experimental/host/netlink/onewire.go

```diff
@@ -0,0 +1,511 @@
+// Copyright 2019 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package netlink
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"sync"
+	"syscall"
+
+	"periph.io/x/periph"
+	"periph.io/x/periph/conn/onewire"
+	"periph.io/x/periph/conn/onewire/onewirereg"
+)
+
+// New opens a 1-wire bus via its netlink interface as described at
+// https://www.kernel.org/doc/Documentation/w1/w1.netlink
+//
+// masterID is the bus number reported by the netlink W1_LIST_MASTERS command.
+// The resulting object is safe for concurrent use.
+//
+// NOTE: the Linux 1-wire netlink API does not support strong pull-ups after
+// write operations. Hence this driver does not support this feature either. The
+// pull-up argument passed to Tx() is ignored. Devices may need to be powered
+// externally to work with this driver.
+func New(masterID uint32) (*OneWire, error) {
+	if isLinux {
+		return newOneWire(masterID)
+	}
+	return nil, errors.New("netlink-onewire: is not supported on this platform")
+}
+
+// OneWire is a 1-wire bus via netlink.
+//
+// It can be used to communicate with multiple devices from multiple goroutines.
+type OneWire struct {
+	masterID uint32
+
+	mu  sync.Mutex
+	s   *socket
+	seq uint32
+}
+
+// Close closes the handle to the 1-wire driver. It is not a requirement to
+// close before process termination.
+func (o *OneWire) Close() error {
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if err := o.s.close(); err != nil {
+		return fmt.Errorf("netlink-onewire: %v", err)
+	}
+	return nil
+}
+
+// String returns the name of the OneWire instance.
+func (o *OneWire) String() string {
+	return fmt.Sprintf("Netlink-OneWire%d", o.masterID)
+}
+
+// Tx performs reset, write and (if len(r) > 0) read operations on the 1-wire
+// bus.
+//
+// NOTE: the Linux 1-wire netlink API does not support requesting strong
+// pull-ups after write operations. Hence this driver does not support this
+// feature either. The pull-up argument passed to Tx() is ignored.
+func (o *OneWire) Tx(w, r []byte, _ onewire.Pullup) error {
+	// Grouping multiple commands into a single netlink message appears to make
+	// bus transactions significantly more stable.
+	cmds := []*w1Cmd{w1CmdReset(), w1CmdWrite(w)}
+	if l := len(r); l > 0 {
+		cmds = append(cmds, w1CmdRead(l))
+	}
+	m := &w1Msg{
+		typ:      msgMasterCmd,
+		masterID: o.masterID,
+		cmds:     cmds,
+	}
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.seq++
+
+	d, err := o.s.sendAndRecv(o.seq, m)
+	if err != nil {
+		return fmt.Errorf("netlink-onewire: transaction failed: %v", err)
+	}
+	copy(r, d)
```

Comment:
I don't understand why this line.

Reply:
Copying to r is correct: d is the data read from the netlink socket (after removing all headers), r is the read buffer passed to Tx() that needs to be populated.

Reply:
Ugh, brain fart.